akka-discovery-azure-api/src/main/scala/akka/discovery/azureapi/AzureRbacAKSServiceDiscovery.scala

```diff
@@ -0,0 +1,270 @@
+import akka.discovery.azureapi.JsonFormat._
+import akka.discovery.{ Lookup, ServiceDiscovery }
+import akka.dispatch.Dispatchers.DefaultBlockingDispatcherId
+import akka.event.Logging
+import akka.http.scaladsl.model.headers.{ Authorization, OAuth2BearerToken }
+import akka.http.scaladsl.model.{ HttpRequest, StatusCodes, Uri }
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.http.scaladsl.{ ConnectionContext, Http, HttpsConnectionContext }
+import akka.pki.kubernetes.PemManagersProvider
+import com.azure.core.credential.{ AccessToken, TokenRequestContext }
+import com.azure.identity.DefaultAzureCredentialBuilder
+
+import java.net.InetAddress
+import java.nio.file.{ Files, Paths }
+import java.security.{ KeyStore, SecureRandom }
+import javax.net.ssl.{ KeyManager, KeyManagerFactory, SSLContext, TrustManager }
+import scala.collection.immutable
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.jdk.FutureConverters._
+import scala.util.Try
+import scala.util.control.{ NoStackTrace, NonFatal }
+
+object AzureRbacAKSServiceDiscovery {
+  private def azureDefaultCredential =
+    new DefaultAzureCredentialBuilder().build()
+
+  private val accessTokenRequestContext = new TokenRequestContext()
+
+  /**
+   * INTERNAL API
+   *
+   * Finds relevant targets given a pod list. Note that this doesn't filter by name as it is the job of the selector
+   * to do that.
+   */
+  @InternalApi
+  private[azureapi] def targets(
+      podList: PodList,
+      portName: Option[String],
+      podNamespace: String,
+      podDomain: String,
+      rawIp: Boolean,
+      containerName: Option[String]): immutable.Seq[ResolvedTarget] =
+    for {
+      item <- podList.items
+      if item.metadata.flatMap(_.deletionTimestamp).isEmpty
+      itemSpec <- item.spec.toSeq
+      itemStatus <- item.status.toSeq
+      if itemStatus.phase.contains("Running")
+      if containerName.forall(name =>
+        itemStatus.containerStatuses match {
+          case Some(statuses) => statuses.filter(_.name == name).exists(!_.state.contains("waiting"))
+          case None           => false
+        })
+      ip <- itemStatus.podIP.toSeq
+      // Maybe port is an Option of a port, and will be None if no portName was requested
+      maybePort <- portName match {
+        case None =>
+          List(None)
+        case Some(name) =>
+          for {
+            container <- itemSpec.containers
+            ports <- container.ports.toSeq
+            port <- ports
+            if port.name.contains(name)
+          } yield Some(port.containerPort)
+      }
+    } yield {
+      val hostOrIp = if (rawIp) ip else s"${ip.replace('.', '-')}.$podNamespace.pod.$podDomain"
+      ResolvedTarget(
+        host = hostOrIp,
+        port = maybePort,
+        address = Some(InetAddress.getByName(ip))
+      )
+    }
```

Comment:
Would be nice if you could split this huge for comprehension into somewhat more digestible pieces, not exactly sure how, maybe some helper methods like `runningPods(...), portFor(...)` to get a bit higher level operations you can then use in a for comprehension?